clippy_lints/src/methods/owned_to_owned.rs

```diff
@@ -0,0 +1,43 @@
+use crate::utils::{is_copy, match_def_path, match_type, span_lint_and_sugg, sugg};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::ExprKind;
+use rustc_lint::LateContext;
+use rustc_middle::ty::TyS;
+
+use super::OWNED_TO_OWNED;
+
+pub fn lint(cx: &LateContext<'_>, expr: &hir::Expr<'_>, expected_path: &[&str], expected_type: Option<&[&str]>) {
+    if_chain! {
+        if let ExprKind::MethodCall(method_path, _, [arg], _) = &expr.kind;
+        if let Some(meth_did) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
+        if let Some(snippet) = sugg::Sugg::hir_opt(cx, arg);
+        let return_type = cx.typeck_results().expr_ty(&expr);
+        let input_type = cx.typeck_results().expr_ty(arg);
+        if TyS::same_type(return_type, input_type);
+        if match_def_path(cx, meth_did, expected_path) &&
+            expected_type.map_or(true, |expected_type_path| match_type(cx,input_type,expected_type_path));
```

Comment:
I believe the "expected type" check is redundant since you are verifying the method.

Reply:
For the current version here, I believe you are right but further changes in prep of using diagnostic items instead (including the ones I'm trying to get added in https://github.com/rust-lang/rust/pull/82128) checks _either_ the trait of the method (for `to_owned`) or the return type (for `to_vec`, …)

I can commit/push a new checkpoint if you'd like.

Reply:
I believe this is no longer redundant after some cleanup refactoring + moving to being able to use diagnostic items (including the ones I'm trying to have added in https://github.com/rust-lang/rust/pull/82128)

I haven't committed/pushed that change yet, but I check if the method belongs to a trait_path _or_ the type.

I'll be pushing my new changes once I've gone through everything else.

Reply:
I should clarify: when I say "to be able to use diagnostic items" I mean hopefully easily switch the function from taking paths to the function taking diagnostic items -- not a mix of both.

Reply:
> checks _either_ the trait of the method (for `to_owned`) or the return type (for `to_vec`, …)

I think it would be better to just check the method for every case. If you only check the return type, then you might have a false positive if someone rolls their own trait with `to_vec` with different semantics (even though that would be a little odd). If you just check the method, there should be no such risk.

Reply:
I guess by "check the method" while using diagnostic types, you mean "check the type/trait the method is on as well as checking the diagnostic item of return value of the type itself"?

Also, I just noticed `os_string.to_os_string()` is actually a method on `OsStr` so I will need to make sure that has a diagnostic item as well.

Reply:
"check the method":
1) the method name is "to_vec"
2) the method is associated with type X
3) type X is diagnostic item "slice"

Not sure how to make the "associated" link but maybe `tcx.associated_item` and/or `tcx.impl_of_method`. We might need a util method to make this easier like `is_diagnostic_assoc_fn(def_id, sym::slice, sym::to_vec)`.